### ISSUE TITLE
akka-http: scaladsl.model.Uri.withQuery doesn't URL encode parameters

### ISSUE THREAD

**Issue by abhijitsarkar** [akka/akka#18574]
Repro steps:

```
import akka.http.scaladsl.model.Uri

val uri = Uri("http://awesome.me").withQuery(Map("q" -> "@awesome"))

assert(uri.toString == "http://awesome.me?q=%40awesome")

```

Actual `http://awesome.me?q=@awesome`

**Comment by jeffrey-aguilera** [comment]
@ is a pchar and does not need to be encoded in the query string.

https://tools.ietf.org/html/rfc3986 <https://tools.ietf.org/html/rfc3986> see section 3.3

**Comment by jrudolph** [comment]
I agree with @jeffrey-aguilera, there's no need to percent encode `'@'` in the query part.

**Comment by ktoso** [comment]
Didn't yet get around verifying yet, thanks for having a look! Ticket seems invalid then, closing.

**Comment by abhijitsarkar** [comment]
Twitter, and OAuth in general, would disagree with you. Not encoding `@`only makes sense for a path like `mailto:fred@example.com`. I've not read the full RFC yet but I'd not be surprised if there's a similar statement in there.

**Comment by ktoso** [comment]
Hi @abhijitsarkar,
can you give a specific example where Akka's behaviour is breaking with some existing APIs for example?

The spec defines the `@` char as valid in the *query* part of an URI (quote grammar from https://tools.ietf.org/html/rfc3986#section-3.3 ):

```
      query       = *( pchar / "/" / "?" )
      pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
```

So I agree with previous commenters, that we should not escape it automatically.

Akka HTTP aims to be as close to the specifications as possible. There always are some APIs which do weird things, however instead making the edge cases the core of the impl, we stick to the spec and if needed provide escape hatches if strictness would forbid interacting with some API. 
In this specific case it seems like escaping can easily be done by the user-code, so I'm reluctant to change HTTP unless provided with more data backing need. Thanks in advance for providing more infos!

**Comment by jrudolph** [comment]
Let's not conflate the issues here:
 * There's a specification for URIs which clearly describes the grammar for valid URIs. akka-http tries to follow that grammar and it allows literal '@' characters in the query. That's also the behavior browsers implement.
 * There's the "issue" that a URI allows multiple syntactic representations for the same value (like here where '@' may or may not be percent encoded). So, if you rely on a "canonical representation" like for (OAuth) signing of requests you may have to add additional constraints to make sure the translation is unambiguous.

In fact, [it seems that OAuth signing defines stricter rules for percent encoding than RFC 3986](http://oauth.net/core/1.0/#encoding_parameters) but that's actually not the concern of a general URI infrastructure. AFAICS, both sides should use the same normalization procedure to make sure that they will arrive at the same hashes, so it wouldn't matter if the actual request target URI is in normalized form or not (I only skimmed the OAuth 1 document, so take this with a grain of salt).

In summary, this means that for purposes of signing a request you cannot rely on `Uri.toString` but need to build your own URI serialization logic which you need anyways, because OAuth also relies on a canonical sort order of parameters which `Uri` also doesn't enforce (or knows about).

**Comment by abhijitsarkar** [comment]
@ktoso 
You can find an example in the [scauth](https://github.com/abhijitsarkar/scala/tree/master/scauth) project, which is an OAuth library I'm creating based on Scala and akka-http. The files that're interesting in particular are [OAuthRequest](https://github.com/abhijitsarkar/scala/blob/master/scauth/src/main/scala/name/abhijitsarkar/scala/scauth/model/OAuthRequest.scala) which builds the Uri, [TwitterService](https://github.com/abhijitsarkar/scala/blob/master/scauth/src/test/scala/name/abhijitsarkar/scala/scauth/example/TwitterService.scala) that currently encodes the `@` and [TwitterApp](https://github.com/abhijitsarkar/scala/blob/master/scauth/src/test/scala/name/abhijitsarkar/scala/scauth/example/TwitterApp.scala) that sends the query.

**Comment by abhijitsarkar** [comment]
I don't have a strong feeling for this and I'd not want you to deviate form the spec. Feel free to close the issue.

**Comment by abhijitsarkar** [comment]
I don't have a strong feeling for this and I'd not want you to deviate form the spec. Closing the issue.

**Comment by eugenemiretsky** [comment]
@asarkar: how did you end up solving this?

**Comment by asarkar** [comment]
@eugenemiretsky Can you be more specific? Are you asking how did I solve how to encode `@` or are you asking how did I encode any URL parameters in general?
Also, instead of hijacking this thread that's been closed for almost 2 years, feel free to open a question on my [Akka](https://github.com/asarkar/akka) repo with the clarification.

**Comment by bpholt** [comment]
Since I found this thread while trying to search for a related issue (how to preserve `%3A` in pre-signed S3 URLs), I thought [this documentation of ”synthetic headers“](http://doc.akka.io/docs/akka-http/10.0.4/scala/http/common/http-model.html#synthetic-headers) allowing the caller to set a raw request URI might be useful for the next person who stumbles through here.

**Comment by raboof** [comment]
> difference being that I am looking to redirect a signed url

(this was cross-posted to https://stackoverflow.com/questions/64383102/how-to-prevent-percent-decoding-of-uris-in-akka-http-redirect-requests, answered there)
